rfcs/2021-03-26-6517-end-to-end-acknowledgement.md

```diff
@@ -0,0 +1,279 @@
+# RFC 6517 - 2021-03-36 - End-to-end Acknowledgement
+
+This RFC proposes a framework for tracking the flow of events such that the
+source can acknowledge them only after they are successfully delivered.
+
+## Scope
+
+This RFC will cover:
+
+* The functional requirements to support end-to-end acknowledgement
+
+* The data model changes and potential impacts to support the above
+
+
+This RFC will not cover:
+
+* Any details of changes to individual sources, transforms, or sinks to
+  necessary to support this feature.
+
+## Motivation
+
+Users want assurance that their data pipeline successfully delivers all the
+events that are presented to it. To accomplish this, Vector needs to be able to
+maintain information on each event tying it back to the source that originated
+the event, so that source can correctly acknowledge those incoming events at
+the right time.
+
+## Examples
+
+Vector contains a number of features that can cause complications with a naïve
+implementation. As such, we need to itemize a number of scenarios to ensure our
+proposed solution covers all the edge cases. These will start with the simplest
+scenario and then introduce complicating factors.
+
+### Single source through linear transforms to a single sink
+
+This is the simplest of cases. For each event the source produces, the event is
+only modified on its transport through the system. The set of transforms
+represented here are all those not listed explicitly below. When the
+destination sink finishes handling the event, a notification needs to be
+delivered to the originating source indicating the delivery status. Thus, the
+event needs to contain a reference to the originating source and a unique
+message identifier. Different sources will have varying requirements on the
+message identifier.
+
+### Multiple sources to a single sink
+
+When multiple sources are configured to feed into a single sink, that sink may
+end up producing batches containing events from multiple sources. This does not
+really require any additional data tracking, but it is necessary to model the
+acknowledgements as flowing back to the source independently for each event,
+not at a batch level.
+
+### Single source to multiple sinks
+
+If an event is to be sent to multiple sinks, we want the acknowledgement to
+only be issued once it has been acknowledged by all of the sinks. This will
```

Comment:
So this prompts the question about how this should be configured. Is this a source attribute, where it references a particular sink as being authoritative, or a sink attribute marking it as authoritative for all sources feeding to it? Will this be the only model, where a single acknowledgement finalizes the event thus simplifying the data flow, or will configuration where multiple sinks must finalize the event be possible?

Reply:
From a UX standpoint, a sink-level attribute is the least confusing, especially since pipelines can change over time. I also think we can default sinks that are commonly used as the system-of-record, like the `aws_s3` sink, so it "just works" and doesn't force the decision on the user.

Reply:
I agree it's least confusing. How do we want to handle a situation where some events don't go to the sink marked as authoritative, either through a fixed configuration or through a `route` transform?

Reply:
If it's expected/normal behavior then those events would not be acknowledged and everything proceeds as normal.

Reply:
I think the important thing for now is that our strategy is compatible with this being configurable. We don't need to figure out the details of configuration now since it likely won't be part of the initial feature set.